src/v/cluster/rm_stm.h

```diff
@@ -378,6 +378,35 @@ class rm_stm final : public persisted_stm {
         }
     };
 
+    struct request_id {
+        model::producer_identity pid;
+        int32_t seq;
+
+        auto operator<=>(const request_id&) const = default;
+
+        template<typename H>
+        friend H AbslHashValue(H h, const request_id& bid) {
+            return H::combine(std::move(h), bid.pid, bid.seq);
+        }
+    };
+
+    template<typename Func>
+    auto with_request_lock(request_id rid, Func&& func) {
+        auto lock_it = _request_locks.find(rid);
+        if (lock_it == _request_locks.end()) {
+            lock_it
+              = _request_locks.emplace(rid, ss::make_lw_shared<mutex>()).first;
+        }
+        auto r_lock = lock_it->second;
+        return r_lock->with(std::forward<Func>(func))
+          .finally([this, r_lock, rid]() {
+              if (r_lock->waiters() == 1) {
```

Comment:
If fiber is the last holder the lock this if should be `r_lock.ready()`, isn't it?
Because now `waiters==1` signal that smbd else wait on the semaphore inside mutex.

Reply:
actually it can not be `ready()` as ready check if there are available units. I think we need to compare `r_lock->waiters() == 0`

Reply:
> actually it can not be `ready()` as ready check if there are available units. I think we need to compare `r_lock->waiters() == 0`

As understand it should release lock after func inside `wait()`. So in this finally block, units will be 1 (in exception case too)

Reply:
damn, I was doing it inside func but then it was lost during the PR updates, thanks folks!

Reply:
That's my fault, sorry!